File: Core/src/main/java/com/github/ezauton/core/action/PeriodicAction.java
package com.github.ezauton.core.action;

import com.github.ezauton.core.utils.IClock;
import com.github.ezauton.core.utils.Stopwatch;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

/**
 * An action which runs at recurring intervals 🔁. Will run all {@link Runnable}s sequentially every period timeUnit.
 */
public abstract class PeriodicAction extends BaseAction {

    protected final long periodMillis;
    private final List<Runnable> runnables;
    protected IClock clock;
    protected Stopwatch stopwatch;
    protected boolean periodDelayAfterExecution = false;
    private int timesRun = 0;

    /**
     * An action which runs at recurring intervals 🔁. Will run all {@link Runnable}s sequentially every period timeUnit.
     *
     * @param period
     * @param timeUnit
     * @param runnables
     */
    public PeriodicAction(long period, TimeUnit timeUnit, Runnable... runnables) {
        this.periodMillis = timeUnit.toMillis(period);
        this.runnables = new ArrayList<>(Arrays.asList(runnables));
    }

    /**
     * Creates a PeriodicAction which will run all {@link Runnable}s sequentially every 20 ms.
     *
     * @param runnables
     */
    public PeriodicAction(Runnable... runnables) {
        this(20, TimeUnit.MILLISECONDS, runnables);
    }

    /**
     * Creates a PeriodicAction which will run every 20 ms.
     */
    public PeriodicAction() {
        this(20, TimeUnit.MILLISECONDS);
    }


    /**
     * Add a runnable to tasks which are executed periodically
     *
     * @param runnable
     * @return
     */
    public PeriodicAction addRunnable(Runnable runnable) {
        runnables.add(runnable);
        return this;
    }

    public int getTimesRun() {
        return timesRun;
    }

    /**
     * An alternative to {@link PeriodicAction#addRunnable(Runnable)}.
     *
     * @param updateableFunc
     * @return
     * @see <a href = "https://stackoverflow.com/a/9584671/4889030">https://stackoverflow.com/a/9584671/4889030</a>
     */
    public PeriodicAction addRunnable(Function<PeriodicAction, Runnable> updateableFunc) {
        runnables.add(updateableFunc.apply(this));
        return this;
    }

    /**
     * Called when the periodic action is first initialized
     */
    protected void init() throws Exception  {
    }

    /**
     * Called every period cycle. By default, adds given {@link Runnable}s
     */
    protected void execute() throws Exception {
        runnables.forEach(Runnable::run);
    }

    /**
     * @return If the action is finished. Will stop execution if returns true.
     */
    protected abstract boolean isFinished() throws Exception;

    /**
     * The action will attempt to try to run as close as it can to the given period.
     *
     * @return true if period is calculated after execution or false if the period counts execution time
     */
    public boolean isPeriodDelayAfterExecution() {
        return periodDelayAfterExecution;
    }

    /**
     * The action will attempt to try to run as close as it can to the given period.
     *
     * @param periodDelayAfterExecution true if period is calculated after execution or false if the period counts execution time
     */
    public void setPeriodDelayAfterExecution(boolean periodDelayAfterExecution) {
        this.periodDelayAfterExecution = periodDelayAfterExecution;
    }

    @Override
    public final void run(ActionRunInfo actionRunInfo) throws Exception  {
        this.clock = actionRunInfo.getClock();

        stopwatch = new Stopwatch(clock);
        stopwatch.reset();

        long start = clock.getTime();

        init();
        do {
            execute();
            long afterExecution = clock.getTime();

            long wait;
            if (isPeriodDelayAfterExecution()) {
                wait = periodMillis;
            } else {
                long millisTotal = afterExecution - start;

                timesRun++;

                long expectedNext = periodMillis * timesRun;

                wait = expectedNext - millisTotal;
            }

            try {
                if (wait < 0) {
                    // TODO: probably should be an exception or a better way of displaying than this. (needs to be catchable though)
                    System.out.printf("The action is executing slower than the set period! milliseconds behind: %d\n", -wait);
                } else if (wait > 0) {
                    clock.sleep(wait, TimeUnit.MILLISECONDS);
                }
            } catch (InterruptedException e) {
                return;
            }
        }
        while (!isFinished() && !isStopped());
    }

    /**
     * A stopwatch which returns the time since the action started running (unless popped)
     * TODO Should probably have a stopwatch cannot be reset (which is public) and another one which can
     * TODO (and is protected) due to encapsulation
     *
     * @return
     */
    public final Stopwatch getStopwatch() {
        return stopwatch;
    }

    /**
     * The clock which the action is run by
     *
     * @return
     */
    public final IClock getClock() {
        return clock;
    }

}

File: Core/src/main/kotlin/com/github/ezauton/core/action/PeriodicAction.kt
package com.github.ezauton.core.action

import com.github.ezauton.conversion.Time
import com.github.ezauton.conversion.millis
import com.github.ezauton.conversion.now
import com.github.ezauton.core.action.require.ResourceHold
import com.github.ezauton.core.action.require.combineTake
import com.github.ezauton.core.utils.Stopwatch
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay

enum class ResourceManagement {
  UNTIL_FINISH,
  LET_GO_EACH_CYCLE
}

val DEFAULT_PERIOD = 20.millis
val DEFAULT_RESOURCE_MANAGEMENT = ResourceManagement.UNTIL_FINISH

interface PeriodicScope : CoroutineScope {
  val stopwatch: Stopwatch
  fun stop(): Nothing
  val start: Time
  val iteration: Int
}

class StopException : Throwable()

private class PeriodicScopeImpl(val scope: CoroutineScope) : PeriodicScope, CoroutineScope by scope {
  override var iteration = 0
  override val stopwatch: Stopwatch = Stopwatch.new()

  var shouldStop = false

  init {
    stopwatch.init()
  }

  override val start = now()
  override fun stop(): Nothing {
    throw StopException()
  }
}

enum class DelayType {
  FROM_START,
  FROM_END,
}

data class Periodic(
  val period: Time = DEFAULT_PERIOD,
  val loopMethod: DelayType = DelayType.FROM_START,
  val duration: Time? = null,
  val iterations: Int? = null,
  val before: Action<*>? = null,
  val after: Action<*>? = null,
  val resourceManagement: ResourceManagement = DEFAULT_RESOURCE_MANAGEMENT,
  val resourcePriorities: List<ResourcePriority> = emptyList(),
) {
  companion object {
    val DEFAULT = Periodic()
  }
}

suspend fun <T> periodic(params: Periodic, block: suspend (PeriodicScope) -> T) = coroutineScope {
  periodic(params.period, params.loopMethod, params.duration, params.iterations, params.before, params.after, params.resourceManagement, *params.resourcePriorities.toTypedArray(), block = block)
}

fun <T> periodicAction(
  period: Time = DEFAULT_PERIOD,
  loopMethod: DelayType = DelayType.FROM_START,
  duration: Time? = null,
  iterations: Int? = null,
  before: Action<*>? = null,
  after: Action<*>? = null,
  resourceManagement: ResourceManagement = DEFAULT_RESOURCE_MANAGEMENT,
  vararg resourcePriorities: ResourcePriority,
  block: suspend (PeriodicScope) -> T
): Action<List<T>> = action {
  return@action periodic(period, loopMethod, duration, iterations, before, after, resourceManagement, *resourcePriorities) {
    block(it)
  }
}


suspend fun <T> periodic(
  period: Time = DEFAULT_PERIOD,
  loopMethod: DelayType = DelayType.FROM_START,
  duration: Time? = null,
  iterations: Int? = null,
  before: Action<*>? = null,
  after: Action<*>? = null,
  resourceManagement: ResourceManagement = DEFAULT_RESOURCE_MANAGEMENT,
  vararg resourcePriorities: ResourcePriority,
  block: suspend (PeriodicScope) -> T
): List<T> = coroutineScope {

  val list = ArrayList<T>()

  val state = PeriodicScopeImpl(this)

  suspend fun doHold(): ResourceHold {
    return resourcePriorities.toList().combineTake()
  }

  fun waitTime(): Time {
    val afterExecution = now()
    return when (loopMethod) {
      DelayType.FROM_START -> {
        val millisTotal = afterExecution - state.start
        state.iteration++
        val expectedNext = period * state.iteration
        expectedNext - millisTotal
      }
      DelayType.FROM_END -> period
    }
  }

  suspend fun doDelay() {
    val waitMillis = waitTime().millisL
    if (waitMillis < 0) {
      println("waitMillis neg 0 no delay")
      System.out.printf("The action is executing slower than the set period! milliseconds behind: %d\n", -waitMillis)
    } else if (waitMillis > 0) {
      delay(waitMillis)
    }
  }

  val stopwatch = Stopwatch.new().reset()

  fun isFinished(): Boolean {
    if (iterations != null && iterations == state.iteration) return true
    if (duration != null && duration <= stopwatch.read()) return true
    return false
  }

  suspend fun run(): T {
    before?.run()
    val res = block(state)
    after?.run()
    return res
  }

  suspend fun letGoEachCycle() {
    val held = doHold()
    var ranOnce = false
    do {
      if (!ranOnce) {
        ranOnce = true
      } else held.giveBack()
      val res = run()
      list.add(res)
      held.giveBack()
      try {
        doDelay()
      } catch (e: CancellationException) {
        held.giveBack()
        return
      }
    } while (!isFinished())
  }

  suspend fun untilFinish() {
    val held = doHold()
    do {
      val res = run()
      list.add(res)
      try {
        doDelay()
      } catch (e: CancellationException) {
        return
      } finally {
        held.giveBack()
      }
    } while (!isFinished())
  }

  try {
    when (resourceManagement) {
      ResourceManagement.LET_GO_EACH_CYCLE -> letGoEachCycle()
      ResourceManagement.UNTIL_FINISH -> untilFinish()
    }
  } catch (e: StopException) {
    // disregard
  }

  return@coroutineScope list

}
